### DISCUSSION TITLE
How to Add Module Component to Modules and Represent Daughter Cards

### DISCUSSION THREAD

**Discussion by tariqwaseem** [netbox-community/netbox#12211] [Help Wanted!]
How in NetBox we can add module component to modules and represent daughter cards just like installing child devices in devices bays of parent devices as I could not find any way to do so; however, there is workaround to define such module as child device which then can accept installation of daughter card/module but we have redundant devices added to the device count which are actually modules to accommodate installation modules within modules.

**Comment by candlerb** [comment]
Child devices and device bays are *never* meant to be used for items which extend a device (e.g. line cards). They are for independent devices like blade servers, or for physical partitioning of some rack space to take smaller-sized devices. See the warning notes [here](https://docs.netbox.dev/en/stable/models/dcim/devicebay/) and [here](https://docs.netbox.dev/en/stable/models/dcim/devicetype/).

Modules are what you should use for things like line cards. Module "components" means ports which the module provides to the device (e.g. interfaces, console ports, console server ports etc)

Modules aren't hierarchical, but you can also attach "inventory items" to individual ports; these can be used to record things like SFPs.

If you can be more specific about the actual product/vendor device that you are trying to model, and what its options and modules are, we might be able to suggest the best way to model it in Netbox.

**Reply by tariqwaseem** [reply]
Thanks dear @candlerb 
I understand the concept of device bays and parent/child relation but I have mentioned them as a reference relation & functionality that should also be available for modules to insert daughter/sub-modules on main module which offer additional interfaces to parent/main module.
For instance, Huawei SmartAX 5600/5603 DSLAM has such control module; namely, SCUB, which has two slots to provide GE electrical or Optical interfaces by installing daughter cards O2GS for 2xGE optical in one slot on module and/or O4GS or E4GFA daughter card on 2nd slot either for 4xGE optical or 4xGE electrical interfaces. Without these expansion cards there is no interfaces available on the SCUB module apart from Console, LCT and MON/TL1 interfaces.
In this scenario, I could not find anyway to add interfaces without adding sub-modules whose configuration varies site-wise. However, I think of two scenarios right now as appended below:
1. Declare corresponding module bay as device bay and define SCUB module as child device to hold O2GS, O4GS or E4GFA sub-modules but it adds redundant module devices.
2. Or, I can add interfaces directly to SCUB without going into sub-modules without mentioning sub-modules and modify interface types on the go site wise.

Please guide me if there is other more convenient option available or this requirement could qualify for new feature requirement?

**Reply by candlerb** [reply]
Definitely *don't* do scenario 1: I have been burnt this way before. The child card will be treated as an independent "device" in the listing of devices; you won't be able to see interfaces on this card or connections to other devices when viewing the main device; you won't be able to do LAGs between interfaces on this card and other cards. Trust me, it's not good.

Creating the SCUB module type with no interfaces, and then adding the interfaces by hand afterwards, sounds like a reasonable approach.

Another option is having a SCUB module with particular GE or optical interfaces "pre-installed" in the module type.  You can then either modify the interfaces if they are the wrong types; or you can define multiple version of the SCUB module type with different combinations of interfaces.

Another option would be to have:
- the SCUB module itself as an "inventory item" (i.e. it doesn't consume a module slot)
- the module bays in the parent device take the daughter cards directly
It feels like a less accurate physical modeling, but it may be acceptable if there are many types and combinations of daughter cards that you have to deal with.

**Reply by tariqwaseem** [reply]
Dear @candlerb 
For me right now, your suggestion is very helpful as workaround to have multiple instances of SCUB with different configurations instead of using module as device. Thanks in this regard.
Should I mark the discussion answered or any thing else we can do to improve the functionality?

**Reply by yousefii** [reply]
Hello dear @candlerb
I have almost the same problem here with a small difference, my main Module (Huawei LPUF-120 )has 2 subslot and also I have 2 daughter boards  that each of them has 6 interfaces (totally I have 12 interfaces in my slot)
I added the Huawei LPUF-120 module as an inventory item to my device but the main problem is that I want to add 2 submodules to my first Router's slot but I can add just one of the daughter boards in each slot.

**Reply by candlerb** [reply]
The solutions you could consider are:
1. Three module bays in the parent device. One contains the LPUF-120 module, and the other two contain the daughter modules. That is: you "flatten" the hierarchy, so all modules are "inserted" directly into the parent device.
2. A single module bay, and a single LPUF-120 module type, which has 12 interfaces.  Make different versions of the LPUF-120 module type to represent different combinations of daughter card(s) inserted into the module.
3. A single module bay, and single LPUF-120 module type, which has no interfaces. Manually add the interfaces to the module when a daughter card is inserted.  Add the daughter cards as inventory items to the parent device.
4. Don't use modules at all, and use Inventory Items instead. These have a hierarchical capability (each can have a "parent" inventory item), but don't have interfaces. So you'd also have to manually add the interfaces, as in option (3).

Option (1) best represents the capabilities of each module, i.e. which interfaces are on which module, and gives the maximum flexibility for "mix and match" insertion of different combinations of modules. If in future modules gain the ability to set a "parent" module, it would also be easy to modify existing data. I think this is what I'd choose today.

Option (4) best represents the hierarchical relationship of modules within the constraints of the current data model, but completely loses the relationship between interfaces and modules.

**Reply by yousefii** [reply]
Thanks for your support @candlerb 
I make three modules bays:
Slot 1 ---> LPUF-120
Slot 1/0 ---> Daughter board#1
Slot 1/1 ---> Daughter board#2
 and add 6 Interfaces on each daughter board with a specific slot number, for example :
for Daughter board#1 : GigabitEthernet1/0/[0-5]
for Daughter board#2 : GigabitEthernet1/1/[0-5]
it seems a good solution but I have to make interfaces manually for every daughter board in each device.

**Reply by candlerb** [reply]
> it seems a good solution but I have to make interfaces manually for every daughter board in each device.

That shouldn't be necessary. You can create components (i.e. interfaces) on the "Module Type" for the daughter board, and then every time you create an instance of that module, it will copy the components from the module type.

View the Module Type, and then use the Add Components menu there.

However, if you've already created the Modules, then adding interfaces to Module Type will not back-fill them to the Module instances. You'd need to script that. (I have one for [copying Device components from Device Type](https://github.com/netbox-community/customizations/blob/master/scripts/add_device_type_components.py), but not Module components from Module Type)

**Reply by candlerb** [reply]
Ah sorry, maybe the problem is to do with correctly *naming* the interfaces. If so then you're probably stuck, although I think you should be able to do bulk *rename* of interfaces after creating a module instance (I've not tried this scenario exactly, but bulk renaming of "normal" interfaces does work)

**Reply by yousefii** [reply]
@candlerb 
finally, I made it!
I use this script for adding interfaces to daughter boards:
GigabitEthernet{module}/[0-5]
Now depending on inserting the daughter board into the slot, the bay number will be replaced with {module}
thank you very much for your support

**Reply by aquirogamon** [reply]
Hello, can you please send an example?
